File: jscomp/others/belt_MutableSet.ml

(* Copyright (C) 2017 Hongbo Zhang, Authors of ReScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module Int = Belt_MutableSetInt
module String = Belt_MutableSetString

module N = Belt_internalAVLset
module A = Belt_Array
module Sort = Belt_SortArray


type ('k, 'id) id = ('k, 'id) Belt_Id.comparable
type ('key, 'id ) cmp = ('key, 'id) Belt_Id.cmp


type ('value,'id) t =
  {
    cmp: ('value, 'id) cmp;
    mutable data: 'value N.t
  } 



let rec remove0 nt x ~cmp = 
  let k = nt.N.value in 
  let c = cmp x k [@bs] in 
  if c = 0 then 
    let {N.left = l; right = r} = nt in       
    match l, r with 
    | None, _ -> r 
    | _, None -> l 
    | Some _,  Some nr ->  
      nt.right <- (N.removeMinAuxWithRootMutate nt nr);
      Some (N.balMutate nt)
   else 
    begin 
      if c < 0 then 
        match nt.left with         
        | None -> Some nt 
        | Some l ->
          nt.left <- (remove0 ~cmp l x );
          Some (N.balMutate nt)
      else 
        match nt.right with 
        | None -> Some nt 
        | Some r -> 
          nt.right <- (remove0 ~cmp r x);
          Some (N.balMutate nt)
    end

let remove  d  v =  
  let oldRoot = d.data in 
  match oldRoot with 
  | None -> ()
  | Some oldRoot2 ->
    let newRoot = remove0 ~cmp:(Belt_Id.getCmpInternal d.cmp) oldRoot2 v in 
    if newRoot != oldRoot then 
      d.data <- newRoot    


let rec removeMany0 t xs i len ~cmp  =  
  if i < len then 
    let ele = A.getUnsafe xs i in 
    let u = remove0 t ele ~cmp in 
    match u with 
    | None -> None
    | Some t -> removeMany0 t xs (i+1) len ~cmp 
  else Some t    

let removeMany d xs =  
  let oldRoot = d.data in 
  match oldRoot with 
  | None -> ()
  | Some nt -> 
    let len = A.length xs in 
    d.data <- 
      (removeMany0 nt xs 0 len 
        ~cmp:(Belt_Id.getCmpInternal d.cmp))


let rec removeCheck0  nt x removed ~cmp= 
  let k = nt.N.value in 
  let c = (Belt_Id.getCmpInternal cmp) x k [@bs] in 
  if c = 0 then 
    let () = removed .contents<- true in  
    let {N.left = l; right = r} = nt in       
    match l, r with 
    | None, _ -> r 
    | _, None -> l  
    | Some _,  Some nr ->  
      nt.right <- (N.removeMinAuxWithRootMutate nt nr);
      Some (N.balMutate nt)
  else 
    begin 
      if c < 0 then 
        match nt.left with         
        | None -> Some nt 
        | Some l ->
          nt.left <- (removeCheck0 ~cmp l x removed);
          Some (N.balMutate nt)
      else 
        match nt.right with 
        | None -> Some nt 
        | Some r -> 
          nt.right <- (removeCheck0 ~cmp r x removed);
          Some (N.balMutate nt)
    end



let removeCheck d v =  
  let oldRoot = d.data in 
  match oldRoot with 
  | None -> false 
  | Some oldRoot2 ->
    let removed = ref false in 
    let newRoot = removeCheck0 ~cmp:d.cmp oldRoot2 v removed in 
    if newRoot != oldRoot then  
      d.data <- newRoot ;   
    removed.contents



let rec addCheck0  t x added ~cmp  =   
  match t with 
  | None -> 
    added .contents<- true;
    N.singleton x 
  | Some nt -> 
    let k = nt.N.value in 
    let c = cmp x k [@bs] in  
    if c = 0 then t 
    else
      let {N.left = l; right = r} = nt in 
      (if c < 0 then                   
         let ll = addCheck0 ~cmp l x added in
         nt.left <- ll
       else   
         nt.right <- (addCheck0 ~cmp r x added );
      );
      Some (N.balMutate nt)

let addCheck m e = 
  let oldRoot = m.data in 
  let added = ref false in 
  let newRoot = addCheck0 ~cmp:(Belt_Id.getCmpInternal m.cmp) oldRoot e added in 
  if newRoot != oldRoot then 
    m.data <- newRoot;
  added.contents

let add m e = 
  let oldRoot = m.data  in 
  let newRoot = N.addMutate ~cmp:m.cmp oldRoot e  in 
  if newRoot != oldRoot then 
    m.data <- newRoot

let addArrayMutate t xs ~cmp =     
  let v = ref t in 
  for i = 0 to A.length xs - 1 do 
    v .contents<- N.addMutate v.contents (A.getUnsafe xs i)  ~cmp
  done; 
  v.contents
    
let mergeMany d xs =   
  d.data <- (addArrayMutate (d.data) xs ~cmp:d.cmp)


let make (type value) (type identity) ~(id : (value, identity) id) =
  let module M = (val id) in 
  {cmp = M.cmp ; data = None}
    
let isEmpty d = 
  N.isEmpty (d.data)
    
let minimum d = 
  N.minimum (d.data)    
let minUndefined d =
  N.minUndefined (d.data)
let maximum d = 
  N.maximum (d.data)
let maxUndefined d =
  N.maxUndefined (d.data)

let forEachU d f = N.forEachU (d.data) f
let forEach d f = forEachU d (fun[@bs] a -> f a)
let reduceU d acc cb = N.reduceU (d.data) acc cb
let reduce d acc cb = reduceU d acc (fun[@bs] a b -> cb a b)    
let everyU d p = N.everyU (d.data) p
let every d p = everyU d (fun[@bs] a -> p a)    
let someU d p = N.someU (d.data) p
let some d p = someU d (fun[@bs] a -> p a)    
let size d = 
  N.size (d.data)
let toList d =
  N.toList (d.data)
let toArray d = 
  N.toArray (d.data)

let fromSortedArrayUnsafe (type value) (type identity) xs ~(id : (value,identity) id) : _ t =
  let module M = (val id) in 
  {data = (N.fromSortedArrayUnsafe xs); cmp = M.cmp}

let checkInvariantInternal d = 
  N.checkInvariantInternal (d.data)
    
    
let fromArray (type value) (type identity)  data ~(id : (value,identity) id) =
  let module M = (val id) in
  let cmp = M.cmp in 
  {cmp; data = (N.fromArray ~cmp data)}


let cmp d0 d1 = 
  N.cmp ~cmp:(d0.cmp) (d0.data) (d1.data)

let eq d0  d1 = 
  N.eq ~cmp:(d0.cmp) (d0.data) (d1.data)
    
let get d x = 
  N.get ~cmp:d.cmp d.data x
    
let getUndefined  d x = 
  N.getUndefined ~cmp:d.cmp d.data x
    
let getExn d x = 
  N.getExn ~cmp:d.cmp d.data x
    

let split d  key  =     
  let arr = N.toArray d.data in
  let cmp = d.cmp in 
  let i = Sort.binarySearchByU arr key (Belt_Id.getCmpInternal cmp)  in   
  let len = A.length arr in 
  if i < 0 then 
    let next = - i -1 in 
    ({
       data = (N.fromSortedArrayAux arr 0 next);
       cmp}
     ,      
       {data = (N.fromSortedArrayAux arr next (len - next));
        cmp}
    ), false
  else 
    (
       {data = (N.fromSortedArrayAux arr 0 i);
        cmp},
       { data = (N.fromSortedArrayAux arr (i+1) (len - i - 1));
         cmp}
     ), true       

let keepU d p = 
  {data = (N.keepCopyU d.data p );  cmp = d.cmp}

let keep d p = keepU d (fun[@bs] a -> p a)
    
let partitionU d p = 
  let cmp = d.cmp in 
  let a, b = N.partitionCopyU d.data p in 
  {data = a ; cmp}, { data = b; cmp}  

let partition d p = partitionU d (fun[@bs] a -> p a)

let subset a b = 
  N.subset  ~cmp:a.cmp a.data b.data

let intersect a b  : _ t = 
  let cmp = a.cmp in 
  match a.data, b.data with 
  | None, _ -> { cmp; data = None}
  | _, None -> { cmp; data = None}
  | Some dataa0, Some datab0 ->  
    let sizea, sizeb = 
      N.lengthNode dataa0, N.lengthNode datab0 in          
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ; 
    ignore @@ N.fillArray datab0 sizea tmp;
    let p = Belt_Id.getCmpInternal cmp in 
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       || 
       (p 
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0 
       )
    then {cmp; data = None}
    else 
      let tmp2 = A.makeUninitializedUnsafe (Pervasives.min sizea sizeb) in 
      let k = Sort.intersectU tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      {data = (N.fromSortedArrayAux tmp2 0 k);
       cmp}
        
let diff a b : _ t = 
  let cmp = a.cmp in 
  let dataa = a.data in 
  match dataa, b.data with 
  | None, _ -> {cmp; data = None}
  | _, None -> 
    {data = (N.copy dataa); cmp}
  | Some dataa0, Some datab0
    -> 
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in  
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ; 
    ignore @@ N.fillArray datab0 sizea tmp;
    let p = Belt_Id.getCmpInternal cmp in 
    if (p (A.getUnsafe tmp (sizea - 1))
          (A.getUnsafe tmp sizea) [@bs] < 0)
       || 
       (p 
          (A.getUnsafe tmp (totalSize - 1))
          (A.getUnsafe tmp 0) [@bs] < 0 
       )
    then {data = (N.copy dataa); cmp}
    else 
      let tmp2 = A.makeUninitializedUnsafe sizea in 
      let k = Sort.diffU tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      {data = (N.fromSortedArrayAux tmp2 0 k); cmp}

let union a b = 
  let cmp = a.cmp in 
  let dataa, datab =  a.data, b.data  in 
  match dataa, datab with 
  | None, _ -> {data = (N.copy datab); cmp}
  | _, None -> {data = (N.copy dataa); cmp}
  | Some dataa0, Some datab0 
    -> 
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in 
    let totalSize = sizea + sizeb in 
    let tmp = A.makeUninitializedUnsafe totalSize in 
    ignore @@ N.fillArray dataa0 0 tmp ;
    ignore @@ N.fillArray datab0 sizea tmp ;
    let p = (Belt_Id.getCmpInternal cmp)  in 
    if p
        (A.getUnsafe tmp (sizea - 1))
        (A.getUnsafe tmp sizea) [@bs] < 0 then 
        { data = (N.fromSortedArrayAux tmp 0 totalSize); cmp}
    else   
      let tmp2 = A.makeUninitializedUnsafe totalSize in 
      let k = Sort.unionU tmp 0 sizea tmp sizea sizeb tmp2 0 p in 
      {data = (N.fromSortedArrayAux tmp2 0 k); cmp}
      
let has d x =
  N.has ~cmp:d.cmp d.data x

let copy d = {data = (N.copy d.data); cmp = d.cmp}

File: lib/js/Belt_MutableSet.js
'use strict';

let Primitive_int = require("./Primitive_int.js");
let Belt_SortArray = require("./Belt_SortArray.js");
let Belt_internalAVLset = require("./Belt_internalAVLset.js");

function remove0(nt, x, cmp) {
  let k = nt.v;
  let c = cmp(x, k);
  if (c === 0) {
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = remove0(l$1, x, cmp);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = remove0(r$1, x, cmp);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, v) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let newRoot = remove0(oldRoot, v, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function removeMany0(_t, xs, _i, len, cmp) {
  while (true) {
    let i = _i;
    let t = _t;
    if (i >= len) {
      return t;
    }
    let ele = xs[i];
    let u = remove0(t, ele, cmp);
    if (u === undefined) {
      return;
    }
    _i = i + 1 | 0;
    _t = u;
    continue;
  };
}

function removeMany(d, xs) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let len = xs.length;
  d.data = removeMany0(oldRoot, xs, 0, len, d.cmp);
}

function removeCheck0(nt, x, removed, cmp) {
  let k = nt.v;
  let c = cmp(x, k);
  if (c === 0) {
    removed.contents = true;
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeCheck0(l$1, x, removed, cmp);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeCheck0(r$1, x, removed, cmp);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function removeCheck(d, v) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return false;
  }
  let removed = {
    contents: false
  };
  let newRoot = removeCheck0(oldRoot, v, removed, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
  }
  return removed.contents;
}

function addCheck0(t, x, added, cmp) {
  if (t !== undefined) {
    let k = t.v;
    let c = cmp(x, k);
    if (c === 0) {
      return t;
    }
    let l = t.l;
    let r = t.r;
    if (c < 0) {
      let ll = addCheck0(l, x, added, cmp);
      t.l = ll;
    } else {
      t.r = addCheck0(r, x, added, cmp);
    }
    return Belt_internalAVLset.balMutate(t);
  }
  added.contents = true;
  return Belt_internalAVLset.singleton(x);
}

function addCheck(m, e) {
  let oldRoot = m.data;
  let added = {
    contents: false
  };
  let newRoot = addCheck0(oldRoot, e, added, m.cmp);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
  }
  return added.contents;
}

function add(m, e) {
  let oldRoot = m.data;
  let newRoot = Belt_internalAVLset.addMutate(m.cmp, oldRoot, e);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
    return;
  }
  
}

function addArrayMutate(t, xs, cmp) {
  let v = t;
  for (let i = 0, i_finish = xs.length; i < i_finish; ++i) {
    v = Belt_internalAVLset.addMutate(cmp, v, xs[i]);
  }
  return v;
}

function mergeMany(d, xs) {
  d.data = addArrayMutate(d.data, xs, d.cmp);
}

function make(id) {
  return {
    cmp: id.cmp,
    data: undefined
  };
}

function isEmpty(d) {
  let n = d.data;
  return n === undefined;
}

function minimum(d) {
  return Belt_internalAVLset.minimum(d.data);
}

function minUndefined(d) {
  return Belt_internalAVLset.minUndefined(d.data);
}

function maximum(d) {
  return Belt_internalAVLset.maximum(d.data);
}

function maxUndefined(d) {
  return Belt_internalAVLset.maxUndefined(d.data);
}

function forEach(d, f) {
  Belt_internalAVLset.forEach(d.data, f);
}

function reduce(d, acc, cb) {
  return Belt_internalAVLset.reduce(d.data, acc, cb);
}

function every(d, p) {
  return Belt_internalAVLset.every(d.data, p);
}

function some(d, p) {
  return Belt_internalAVLset.some(d.data, p);
}

function size(d) {
  return Belt_internalAVLset.size(d.data);
}

function toList(d) {
  return Belt_internalAVLset.toList(d.data);
}

function toArray(d) {
  return Belt_internalAVLset.toArray(d.data);
}

function fromSortedArrayUnsafe(xs, id) {
  return {
    cmp: id.cmp,
    data: Belt_internalAVLset.fromSortedArrayUnsafe(xs)
  };
}

function checkInvariantInternal(d) {
  Belt_internalAVLset.checkInvariantInternal(d.data);
}

function fromArray(data, id) {
  let cmp = id.cmp;
  return {
    cmp: cmp,
    data: Belt_internalAVLset.fromArray(data, cmp)
  };
}

function cmp(d0, d1) {
  return Belt_internalAVLset.cmp(d0.data, d1.data, d0.cmp);
}

function eq(d0, d1) {
  return Belt_internalAVLset.eq(d0.data, d1.data, d0.cmp);
}

function get(d, x) {
  return Belt_internalAVLset.get(d.data, x, d.cmp);
}

function getUndefined(d, x) {
  return Belt_internalAVLset.getUndefined(d.data, x, d.cmp);
}

function getExn(d, x) {
  return Belt_internalAVLset.getExn(d.data, x, d.cmp);
}

function split(d, key) {
  let arr = Belt_internalAVLset.toArray(d.data);
  let cmp = d.cmp;
  let i = Belt_SortArray.binarySearchBy(arr, key, cmp);
  let len = arr.length;
  if (i >= 0) {
    return [
      [
        {
          cmp: cmp,
          data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, i)
        },
        {
          cmp: cmp,
          data: Belt_internalAVLset.fromSortedArrayAux(arr, i + 1 | 0, (len - i | 0) - 1 | 0)
        }
      ],
      true
    ];
  }
  let next = (-i | 0) - 1 | 0;
  return [
    [
      {
        cmp: cmp,
        data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, next)
      },
      {
        cmp: cmp,
        data: Belt_internalAVLset.fromSortedArrayAux(arr, next, len - next | 0)
      }
    ],
    false
  ];
}

function keep(d, p) {
  return {
    cmp: d.cmp,
    data: Belt_internalAVLset.keepCopy(d.data, p)
  };
}

function partition(d, p) {
  let cmp = d.cmp;
  let match = Belt_internalAVLset.partitionCopy(d.data, p);
  return [
    {
      cmp: cmp,
      data: match[0]
    },
    {
      cmp: cmp,
      data: match[1]
    }
  ];
}

function subset(a, b) {
  return Belt_internalAVLset.subset(a.data, b.data, a.cmp);
}

function intersect(a, b) {
  let cmp = a.cmp;
  let match = a.data;
  let match$1 = b.data;
  if (match === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  if (match$1 === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(match);
  let sizeb = Belt_internalAVLset.lengthNode(match$1);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(match, 0, tmp);
  Belt_internalAVLset.fillArray(match$1, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  let tmp2 = new Array(Primitive_int.min(sizea, sizeb));
  let k = Belt_SortArray.intersect(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function diff(a, b) {
  let cmp = a.cmp;
  let dataa = a.data;
  let match = b.data;
  if (dataa === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  if (match === undefined) {
    return {
      cmp: cmp,
      data: Belt_internalAVLset.copy(dataa)
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(dataa);
  let sizeb = Belt_internalAVLset.lengthNode(match);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa, 0, tmp);
  Belt_internalAVLset.fillArray(match, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
      cmp: cmp,
      data: Belt_internalAVLset.copy(dataa)
    };
  }
  let tmp2 = new Array(sizea);
  let k = Belt_SortArray.diff(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function union(a, b) {
  let cmp = a.cmp;
  let dataa = a.data;
  let datab = b.data;
  if (dataa === undefined) {
    return {
      cmp: cmp,
      data: Belt_internalAVLset.copy(datab)
    };
  }
  if (datab === undefined) {
    return {
      cmp: cmp,
      data: Belt_internalAVLset.copy(dataa)
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(dataa);
  let sizeb = Belt_internalAVLset.lengthNode(datab);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa, 0, tmp);
  Belt_internalAVLset.fillArray(datab, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0) {
    return {
      cmp: cmp,
      data: Belt_internalAVLset.fromSortedArrayAux(tmp, 0, totalSize)
    };
  }
  let tmp2 = new Array(totalSize);
  let k = Belt_SortArray.union(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function has(d, x) {
  return Belt_internalAVLset.has(d.data, x, d.cmp);
}

function copy(d) {
  return {
    cmp: d.cmp,
    data: Belt_internalAVLset.copy(d.data)
  };
}

let Int;

let $$String;

let forEachU = forEach;

let reduceU = reduce;

let everyU = every;

let someU = some;

let keepU = keep;

let partitionU = partition;

exports.Int = Int;
exports.$$String = $$String;
exports.make = make;
exports.fromArray = fromArray;
exports.fromSortedArrayUnsafe = fromSortedArrayUnsafe;
exports.copy = copy;
exports.isEmpty = isEmpty;
exports.has = has;
exports.add = add;
exports.addCheck = addCheck;
exports.mergeMany = mergeMany;
exports.remove = remove;
exports.removeCheck = removeCheck;
exports.removeMany = removeMany;
exports.union = union;
exports.intersect = intersect;
exports.diff = diff;
exports.subset = subset;
exports.cmp = cmp;
exports.eq = eq;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.everyU = everyU;
exports.every = every;
exports.someU = someU;
exports.some = some;
exports.keepU = keepU;
exports.keep = keep;
exports.partitionU = partitionU;
exports.partition = partition;
exports.size = size;
exports.toList = toList;
exports.toArray = toArray;
exports.minimum = minimum;
exports.minUndefined = minUndefined;
exports.maximum = maximum;
exports.maxUndefined = maxUndefined;
exports.get = get;
exports.getUndefined = getUndefined;
exports.getExn = getExn;
exports.split = split;
exports.checkInvariantInternal = checkInvariantInternal;
/* No side effect */
